### DISCUSSION TITLE
BTT Smart Filament Sensor & Precompiled

### DISCUSSION THREAD

**Discussion by CJfoks** [Jyers/Marlin#1215] [Q&A] [answered]
So today I decided to try this awsome branch of Marlin. Problem is I have BTT Sensor, so I thought that prcompiled firmware is a NO. After diggin config and compiling firmware (A Few Hours Later) I discovered that there is option for turning on Filament sensor in menu. Was my effort pointless? I'm asking in case of future updates.

**Comment by Cal-LF** [comment]
No sir. Filament sensor is in the menu along with the runout length. The BTT Smart sensor has to have an additional setting changed to operate. 1482 #define FILAMENT_MOTION_SENSOR in the config.h and then compiled. The runout length can be 10 to 25mm. Sensor detect level is set to low by default and works with the BTT.
The switch in the menu turns it on and off.

**Reply by AlphaDE** [reply]
@Cal-LF I am just fighting with this, the BTT sensor does not work with my precompiled image (the simple on/off sensor did) and although trying to compile the firmware (using @Jyers sources) myself, I yet did not find the right settings. Runout distance is set to 7.0 as suggest by many tutorials. Nevertheless, whenever trying to print, a runout is reported immediately.

**Reply by Cal-LF** [reply]
@AlphaDE The 7.0 zero setting caused it to trip on my printer but not immediately. Try 10 or more. Another firmware has it set at 25mm (1") and it works fine on mine, just takes longer to trip. 
Another user also found he had his BTT too far from the extruder and caused it to trip. Mine is next to the extruder on a removable mount. 
The plug for the BTT fitted onto the mainboard ok?

**Reply by AlphaDE** [reply]
@Cal-LF Thanks.

The problem on my part seems to be more in the hardware.

I searched around a lot and tried different settings in the firmware.

Then I found a PDF on the BTT sensor on Github stating that the connector cable should be wired G(round, long solid markers), S(ignal, few short markers), V(oltage, many short markers) right to left when looking from the top with the name "Smart Filament Sensor" readable.

My cable there is wired S, G, V, so Signal and Ground are inverted.

I also checked the connector in my Ender 3 V2 4.2.7 motherboard. According to documentation I found that one is wired V, G, S starting from the green hotbed connector block.

The connector there with the supplied cable was wired G, V, S with G and V inverted. The BTT pdf states its V,G,S as it should be

That would have led overall to G was connected to S, S to V and V to G and hopefully not trashing my mainboard.

For the time, I disconnected the sensor and (with the disconnected sensor) suddently got heating errors on bed heating.

For the positioning, as I use direct drive (micro swiss with rails) the sensor is mounted just at the heating box for the filament, but everything connected with tubes so the sensor cannot be pulled by the filament which would cause the wheel not to turn.

**Reply by Cal-LF** [reply]
I would say that reversing the signal and ground would cause no damage. V to ground would cause an issue.

**Reply by AlphaDE** [reply]
In fact the cable was right, the wiring in the YT video of Teaching Tech did not match at least my previous 4.2.2 board. The cable markings in the manual are not used the same way in the shipped product. So I accidently switched G and S, I reverted it, though not sure what's causing bed heating error after printing started even after the sensor was disconnected.
But I found the reason for kill message on bed heating at least, I used the Auto PID function which caused the issue. Resetting to factory settings solved it.

So back to square one, no clue why the firmware does not recognize the sensor. I tried 10 mm run out and will try more.

**Reply by Cal-LF** [reply]
Check that the high - low state is low. High state will give instant trip.

**Reply by AlphaDE** [reply]
```
#define FIL_RUNOUT_STATE     LOW        
#define FIL_RUNOUT_PULLUP
```

**Reply by AlphaDE** [reply]
I just checked the sensor (G vs S) with a multi meter and there while the print is running, there 3.3 V is intermitting with 0 about every 2 secs, so the sensor itself seems to work. I increased to 20 mm which did not help.

**Reply by Cal-LF** [reply]
I think the sensor is suppose to stay at a high state while filament is moving through. If it is jumping high to low every 2 seconds, something is not working properly.
I might be wrong on which state the sensor is suppose to stay.

**Reply by AlphaDE** [reply]
@Cal-LF I'm not sure... from what I read on other places, the sensor is giving and continious "high/low" signal when filament runs through (by the turning disc). Marlin, configured for the motion sensor combined with the runout distance should recognize this and trigger the runout situation if the changing signal is lost (stays either high or low, indicating that the disc is not turning anymore for any reason, like out of filament, blocked nossle or blocked filament) for more than the runout distance.

**Reply by Cal-LF** [reply]
So the firmware decodes the motion with the changing high low signal and a steady state will trip as a runout.

**Reply by AlphaDE** [reply]
That's how I understood it, that keeps the login in the sensor rather easy. The disc triggers a switch and turns it on and off causing the changing signal indication motion. If the motion stops (although the extruder is working) for a lenght of more than runout (to avoid false positives), there is an issue and user action is required.

**Reply by AlphaDE** [reply]
See https://www.youtube.com/watch?v=9gLSJIxiuwo up from 2:20 where the function is explained and shown in a simple logic matching the behaviour of my volts meter.

**Reply by Cal-LF** [reply]
Well, I have no other ideas. The sensor should work with the firmware settings you are using. 
I am testing mriscoc firmware and it is working with the BTT sensor. You could give his a try. Do you have the V427 mainboard?

**Reply by AlphaDE** [reply]
After the current print, I try the HIGH setting and then this firmware. Yes, I upgraded the Ender 3 V2 to a 4.2.7 mainboard. My setup: Ender 3 V2, 4.2.7 mainboard, rails on all axis, MicroSwiss DirectDrive, BLTouch and the BTT, using the Ender 3 V2 LCD display (that's why I came to Jyers firmware as it supports the screen, too) and a Raspberry Pi 4B with OctoPrint.

**Reply by Cal-LF** [reply]
Change .zip to .bin
This does not support the BL Touch.
[Beta-427BTT-ManualMesh-20210715-172642.zip](https://github.com/Jyers/Marlin/files/6878543/Beta-427BTT-ManualMesh-20210715-172642.zip)

**Reply by AlphaDE** [reply]
I just flashed the Jyers firmware with

`#define FIL_RUNOUT_STATE     HIGH`

and

`#define FILAMENT_RUNOUT_DISTANCE_MM 10`

after the last print was finished and started a small print (cover for the barcode for the Ender 3 my nephew will get). 

And guess what, yet, the print does work, no runout message at all. I also disconnected the bowden tube connecting the sensor to the extruder and pulled it out by about 5 cm, so when extruding, the filament would not move through the sensor. And also this worked, during the initial line of the print, I got the "out of filament/change filament" message and the print paused.

So, it needs to be HIGH instead of LOW in FIL_RUNOUT_STATE.

**Comment by AlphaDE** [comment, marked as answer]
Use 

```
#define FIL_RUNOUT_STATE     HIGH
#define FIL_RUNOUT_PULLUP
#define FILAMENT_RUNOUT_DISTANCE_MM 10
```
and uncomment

`#define FILAMENT_MOTION_SENSOR`

Make sure that the wiring is correct, the cables have markings and the assignment on the BTT is (starting at the filament connector):

Ground
Signal
Unassigned
Volt

Check the board connector and reassign if necessary. The supplied cable works for Ender 3 32 bit motherboards (4.2.2 or 4.2.7).

**Reply by Cal-LF** [reply]
I'm glad you got it working. Double checked my trigger state and it is low so maybe there is a difference in our wiring.
Yes only three wires are really used. 
After all that monkey business I bet it is a relief,
